cmd/encryption.go

```diff
@@ -0,0 +1,186 @@
+)
+
+const (
+	SSECustomerAlgorithm = "X-Amz-Server-Side-Encryption-Customer-Algorithm"
+	SSECustomerKey       = "X-Amz-Server-Side-Encryption-Customer-Key"
+	SSECustomerKeyMD5    = "X-Amz-Server-Side-Encryption-Customer-Key-MD5"
+)
+
+// SSECustomerKeySize is the size of valid client provided encryption keys in bytes.
+// Currently AWS supports only AES256. So the SSE-C key size is fixed to 32 bytes.
+const SSECustomerKeySize = 32
+
+// IsSSECustomerRequest returns true if the given HTTP header
+// contains server-side-encryption with customer provided key fields.
+func IsSSECustomerRequest(header http.Header) bool {
+	if header.Get(SSECustomerAlgorithm) != "" {
+		return true
+	}
+	if header.Get(SSECustomerKey) != "" {
+		return true
+	}
+	if header.Get(SSECustomerKeyMD5) != "" {
+		return true
+	}
+	return false
+}
+
+// ParseSSECustomerRequest parses the SSE-C header fields of the provided header.
+// It returns the client provided key on success.
+func ParseSSECustomerRequest(header http.Header) (key []byte, err error) {
+	if algorithm := header.Get(SSECustomerAlgorithm); algorithm != "AES256" { // Currently AWS supports only AES256
+		return nil, errInvalidSSEAlgorithm
+	}
+	if header.Get(SSECustomerKey) == "" {
+		return nil, errMissingSSEKey
+	}
+	if header.Get(SSECustomerKeyMD5) == "" {
+		return nil, errMissingSSEKeyMD5
+	}
+
+	key, err = base64.StdEncoding.DecodeString(header.Get(SSECustomerKey))
+	if err != nil {
+		return nil, errInvalidSSEKey
+	}
+	if len(key) != SSECustomerKeySize {
+		return nil, errInvalidSSEKey
+	}
+
+	keyMD5, err := base64.StdEncoding.DecodeString(header.Get(SSECustomerKeyMD5))
+	if err != nil {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	if md5Sum := md5.Sum(key); !bytes.Equal(md5Sum[:], keyMD5) {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	return key, nil
+}
+
+// RangeWriter returns an io.Writer which ignores all data written
+// to it before start and all data written to it after end.
+func RangeWriter(w io.Writer, start, end int64) io.Writer {
+	if start < 0 {
+		panic("start cannot be negative")
+	}
+	if end < 0 {
+		panic("end cannot be negative")
+	}
+	if start > end {
+		panic("start cannot greater than end")
+	}
+	return &rangeWriter{dst: w, start: start, end: end}
+}
+
+type rangeWriter struct {
+	dst        io.Writer
+	start, end int64
+}
+
+func (w *rangeWriter) Write(p []byte) (n int, err error) {
+	if int64(len(p)) <= w.start {
+		w.start -= int64(len(p))
+		w.end -= int64(len(p))
+		return
+	}
+	if w.start > 0 {
+		n = int(w.start)
+		w.end -= w.start
+		w.start = 0
+		p = p[n:]
+	}
+
+	if int64(len(p)) <= w.end {
+		nn, err := w.dst.Write(p)
+		if err != nil {
+			return n + nn, err
+		}
+		w.end -= int64(nn)
+		return n + nn, err
+	}
+	if w.end > 0 {
+		nn, err := w.dst.Write(p[:w.end])
+		if err != nil {
+			return nn + n, err
+		}
+		p = p[w.end:]
+		w.end = 0
+		n += nn
+	}
+
+	return n + len(p), err
+}
+
+type sseObject struct {
```

Comment:
I guess we don't need to say sseObject instead we can simply say `encryptedObject` since the file is `encryption.go` ?